src/fsm.py

```diff
@@ -0,0 +1,319 @@
+# encoding: utf-8
+
+from state import State, DeadState
+from transition import Transition
+from fsm_exceptions import *
+
+
+class FSM(object):
+
+    def __init__(self):
+        # We are forcing FSM to be "abstract", each FSM instance must be an instance of its subclass
+        assert type(self) is not FSM, 'FSM must be inherited from'
+
+        # Set of all states in this FSM
+        self._states = set()
+
+        # Set of all transitions in this FSM
+        self._transitions = set()
+
+        # Current state of the FSM
+        self._current_state = None
+
+        # Initial state of this FSM
+        self._initial_state = None
+
+        # Set of all symbols
+        self._alphabet = set()
+
+        # When defined, this variable contains the dead state for this FSM.
+        # Note that there will never be a transition that involves a dead state.
+        # Hence, on_transition callback should be merged into on_enter/on_loop_enter callback of the dead state.
+        self._dead_state = None
+
+        # Indicates whether this FSM needs validation before next step
+        self._dirty = False
+
+        # Data structure for finding destination states
+        # (and corresponding callbacks) based on symbol and source state
+        self._map = dict()
+        '''
+        Map is a two-level dict with the following structure:
+        {
+            symbol_1: {
+                src_state_1: (dst_state_1, callback),
+                src_state_2: (dst_state_2, callback),
+                src_state_3: (dst_state_3, callback)
+            },
+            symbol_2: {
+                src_state_1: (dst_state_3, callback),
+                src_state_2: (dst_state_2, callback),
+                src_state_3: (dst_state_1, callback)
+            }
+        }
+        '''
+
+    @property
+    def current_state(self):
```

Comment:
Because it's a library, adding docstrings to everything with parameter and type information would be very nice.